upgrade/upgrade_provider.go

```diff
@@ -366,29 +366,36 @@ func UpgradeProvider(ctx context.Context, repoOrg, repoName string) (err error)
 	}
 
 	if GetContext(ctx).UpgradeBridgeVersion {
-		steps = append(steps, step.Cmd("go", "get",
-			"github.com/pulumi/pulumi-terraform-bridge/v3@"+targetBridgeVersion.String()).
-			In(repo.providerDir()))
-
-		steps = append(steps, step.Cmd("go", "get",
-			"github.com/hashicorp/terraform-plugin-framework").
-			In(repo.providerDir()))
-		steps = append(steps, step.Cmd("go", "get",
-			"github.com/hashicorp/terraform-plugin-mux").
-			In(repo.providerDir()))
-		steps = append(steps, step.Cmd("go", "mod", "tidy").
-			In(repo.providerDir()))
-
-		// Now that we've upgraded the bridge, we want the other places in the repo to use the bridge's
-		// Pulumi version.
-		upgradePulumiEverywhereStep := BridgePulumiVersions(ctx, repo)
-
-		steps = append(steps, upgradePulumiEverywhereStep)
+		steps = append(steps, step.Combined("Upgrade Bridge Version",
+			step.Cmd("go", "get",
+				"github.com/pulumi/pulumi-terraform-bridge/v3@"+targetBridgeVersion.String()),
+			step.Cmd("go", "get", "github.com/hashicorp/terraform-plugin-framework"),
+			step.Cmd("go", "get", "github.com/hashicorp/terraform-plugin-mux"),
+			step.Cmd("go", "mod", "tidy"),
+		).In(repo.providerDir()))
 	}
 	if GetContext(ctx).UpgradePfVersion {
-		steps = append(steps, step.Cmd("go", "get",
-			"github.com/pulumi/pulumi-terraform-bridge/pf@"+targetPfVersion.String()).
-			In(repo.providerDir()))
+		steps = append(steps, step.Combined("Upgrade Pf Version",
+			step.Cmd("go", "get",
+				"github.com/pulumi/pulumi-terraform-bridge/pf@"+targetPfVersion.String()),
+			step.Cmd("go", "mod", "tidy"),
+		).In(repo.providerDir()))
+	}
+
+	if v := GetContext(ctx).TargetPulumiVersion; v != "" {
+		steps = append(steps, step.Combined("Upgrade Pulumi Version",
+			step.Cmd("go", "get", "github.com/pulumi/pulumi/pkg/v3@"+v),
```

Comment:
AS we just discussed and explored, if V is a pre-release such as a hash, we might need to resort to go mod replace, as we're likely not intending to check that in anyway, and we want to make sure that GO does not automatically pick different versions for indirect deps via go mod tidy.

Reply:
Fixed while you were reviewing. See https://github.com/pulumi/upgrade-provider/pull/192/commits/ac333b22f10f97610864bda7535c46ac722c0ea8.